### ISSUE TITLE
Too many queries?

### ISSUE THREAD

**Issue by clytemnestra** [sonata-project/SonataAdminBundle#3922]
The SF debug bar shows too many queries on all my admin classes. My admin classes have nothing custom. I played a bit with my entities and found out that if I delete the ```mappedBy``` fields from my User class, it doesn't do those queries anymore, but then the entity is not valid anymore, since it needs mappedBy for every inversedBy field.

Why am I getting so many queries? Is it normal?

Here's one of my admin classes's example:

![image](https://cloud.githubusercontent.com/assets/7060048/15884782/69bd14cc-2d5b-11e6-8307-09824e97cc48.png)

Here's the query table:
https://jsfiddle.net/nmxz4j7y/1/

The admin's entity

```php
/**
 * @ORM\Table()
 * @ORM\Entity
 */
class BannedUsers
{
    /**
     * @var integer
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @ORM\ManyToOne(targetEntity="User")
     * @ORM\JoinColumn(referencedColumnName="id",onDelete="CASCADE")
     */
    private $user;

    /**
     * @ORM\ManyToOne(targetEntity="User")
     * @ORM\JoinColumn(referencedColumnName="id",onDelete="CASCADE")
     */
    private $censor;

    /**
     * @ORM\Column(type="string",nullable=true,length=2000)
     */
    private $reason;


    /**
     * @ORM\Column(type="boolean")
     */
    private $isBanActive=true;

    /**
     * @ORM\Column(type="datetime")
     * @Timestampable(on="create")
     */
    private $bannedAt;
```

This is my user class, pretty big but maybe it helps:

```php
/**
 * @ORM\Entity(repositoryClass="UserRepository")
 * @ORM\Table(name="fos_user")
 */
class User extends BaseUser
{
    const DEFAULT_PATIENT_CREDIT = 300;

    /**
     * @ORM\Id
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    protected $id;

    /**
     * @ORM\ManyToMany(targetEntity="MedAppBundle\Entity\Group",inversedBy="users")
     * @ORM\JoinTable(name="fos_user_user_group",
     *      joinColumns={@ORM\JoinColumn(name="user_id", referencedColumnName="id",onDelete="CASCADE")},
     *      inverseJoinColumns={@ORM\JoinColumn(name="group_id", referencedColumnName="id")}
     * )
     */
    protected $groups;


    /**
     * @var string
     * @Assert\NotBlank(message="user.firstname.blank",groups={"Profile"})
     * @ORM\Column(name="firstname", type="string",nullable=true)
     */
    protected $firstname;

    /**
     * @var string
     * @Assert\NotBlank(message="user.lastname.blank",groups={"Profile"})
     * @ORM\Column(name="lastname", type="string",nullable=true)
     */
    protected $lastname;

    /**
     * @var string
     *
     * @ORM\Column(name="phoneno", type="string",nullable=true)
     */
    protected $phoneno;

    /**
     * @var string
     *
     * @ORM\Column(name="registration_ip", type="string",nullable=true)
     */
    protected $registrationIp;

    /**
     * @var string
     *
     * @ORM\Column(name="effects", type="boolean")
     */
    protected $effects = true;
    /**
     *
     * @ORM\Column(name="public_email", type="boolean")
     */
    protected $public_email = true;
    /**
     * @Assert\Valid
     * @ORM\OneToOne(targetEntity="MedAppBundle\Entity\MedicSettings",mappedBy="medic", cascade={"persist"})
     * @Assert\Valid()
     */
    protected $medicinfo;
    /**
     * @ORM\OneToMany(targetEntity="MedAppBundle\Entity\MedicComm",mappedBy="medic")
     */
    protected $mediccomm;

    /**
     * @ORM\OneToOne(targetEntity="PrivateMessageBundle\Entity\MessageSettings",mappedBy="user",cascade={"persist"})
     */
    protected $messageSettings;


    /**
     * @ORM\OneToOne(targetEntity="SupportMessageBundle\Entity\TicketSettings",mappedBy="user",cascade={"persist"})
     */
    protected $ticketSettings;

    /**
     * @ORM\OneToOne(targetEntity="Application\Sonata\MediaBundle\Entity\Media",cascade={"persist"})
     * @ORM\JoinColumn(referencedColumnName="id",onDelete="SET NULL")
     */
    protected $profilepic;

    /**
     * @var Media
     *
     * @ORM\OneToMany(targetEntity="Application\Sonata\MediaBundle\Entity\Media", mappedBy="user")
     * @ORM\JoinColumn(name="images", referencedColumnName="id",nullable=true)
     */
    private $images;

    /**
     * @var Media
     *
     * @ORM\OneToMany(targetEntity="Application\Sonata\MediaBundle\Entity\Gallery", mappedBy="user")
     * @ORM\JoinColumn(name="galleries", referencedColumnName="id",nullable=true)
     */
    private $galleries;



    /**
     * @var float
     *
     * @ORM\Column(name="credit", type="float")
     */
    protected $credit = 0;

    /**
     *
     * @ORM\ManyToOne(targetEntity="MedAppBundle\Entity\User")
     * @ORM\JoinColumn(name="referredBy", referencedColumnName="id",nullable=true,onDelete="SET NULL")
     */
    protected $referredBy;

    /**
     * @var string
     *
     * @ORM\Column(name="locationData", type="string",nullable=true)
     */
    protected $locationData;

    /**
     * @ORM\OneToMany(targetEntity="MedAppBundle\Entity\Stream",mappedBy="streamer")
     * @ORM\JoinColumn(referencedColumnName="id")
     */
    private $streams;

    /**
     * @var string
     *
     * @ORM\Column(name="savedFilters", type="string",nullable=true)
     */
    protected $savedFilters;


    /**
     *
     * @ORM\Column(name="registration_date", type="datetime",nullable=true)
     */
    protected $registrationDate;

    /**
     * @ORM\OneToMany(targetEntity="MedAppBundle\Entity\FavoriteUsers",mappedBy="user")
     * @ORM\JoinColumn(referencedColumnName="id")
     */
    private $favUsers;

    /**
     * @var string
     *
     * @ORM\Column(name="chosen_theme", type="string",nullable=true)
     */
    protected $chosenTheme;

    /**
     * @var string
     *
     * @ORM\Column(name="site_lang", type="string",nullable=true)
     */
    protected $siteLang;

    /**
     * @var integer
     *
     * @ORM\Column(name="total_time_live", type="integer", nullable=true)
     *
     */
    protected $totalTimeLive;

    /**
     *
     * @ORM\ManyToOne(targetEntity="FirmBundle\Entity\Firm",inversedBy="firmAdmins", cascade={"persist"})
     */
    protected $adminOfFirm;

    /**
     * @ORM\OneToMany(targetEntity="MedAppBundle\Entity\EmailList",mappedBy="user")
     * @ORM\JoinColumn(referencedColumnName="id")
     */
    private $emailList;

    /**
     * @ORM\OneToOne(targetEntity="MedAppBundle\Entity\Newsletter",mappedBy="user")
     * @ORM\JoinColumn(referencedColumnName="id")
     */
    private $newsletter;

    /**
     * @var string
     *
     * @ORM\Column(name="folderName", type="string", nullable=true)
     */
    public $folderName;

    /**
     *
     * @ORM\Column(name="bonus_received", type="boolean")
     */
    protected $bonusReceived = false;

    /**
     *
     * @ORM\Column(name="purchase_bonus_eligibility", type="boolean", options={"default" = true }, nullable=true )
     */
    protected $purchaseBonusEligibility;

    /**
     * @ORM\OneToMany(targetEntity="MedAppBundle\Entity\PendingProfileImage",mappedBy="user")
     */
    protected $pendingProfileImages;

    /**
     * @ORM\Column(name="is_validated_phone_no", type="boolean", nullable=true)
     */
    private $isValidatedPhoneNo;

    /**
     * @ORM\Column(name="mobile_validation_pin", type="string", nullable=true)
     */
    private $mobileValidationPin;

    /**
     * @ORM\Column(name="username_token", type="string", nullable=true, options={"default" = true })
     */
    private $usernameToken;




    /**
     * @ORM\OneToOne(targetEntity="MedAppBundle\Entity\AccountDocument", mappedBy="user")
     */
    private $accountDocument;


    /**
     * @ORM\OneToMany(targetEntity="MedAppBundle\Entity\AccountDocument", mappedBy="takenByAdmin")
     * @ORM\JoinColumn(referencedColumnName="id",onDelete="CASCADE")
     */
    private $takenPending;
```

If I delete all the mappedBy fields, the same page on my admin shows less queries, as I want it to be, but invalid entity:

![image](https://cloud.githubusercontent.com/assets/7060048/15884903/147cb570-2d5c-11e6-9cb7-93fe621d242e.png)

Here's the queries table
https://jsfiddle.net/nmxz4j7y/

**Comment by tim96** [comment]
@clytemnestra I think you have many queries because of Lazy Load.
I'm not sure but you can try to override method `createQuery` in `Admin` class and join all related data to decrease count query.

```
    public function createQuery($context = 'list')
    {
        $query = parent::createQuery($context);

        $query
            ->addSelect('User')
            ->leftJoin($query->getRootAlias().'.User', 'User')
            ;

        return $query;
    }
```

**Comment by clytemnestra** [comment]
But where or why does the lazy load happen? If I remember correctly, it didn't use to happen before, I'v e just noticed it. 

It is a hassle to override each admin class query, and setting lazy load for each entity is not a good idea, either.

**Comment by OskarStark** [comment]
what about fetch eager on the doctrine mapping?

**Comment by clytemnestra** [comment]
But then every query that I make will do that and I don't want that, it's a bad idea to get all related records everytime if I don't need them(for queries that are outside the admin).

**Comment by tim96** [comment]
> But where or why does the lazy load happen?

it's a very difficult question, but I'm sure it is not sonata admin issue.

**Comment by greg0ire** [comment]
Well I'd say have a look at `createQuery`… BTW, not sure this should be reported here.

**Comment by greg0ire** [comment]
Closing because I can't see how this would not be a Sonata Doctrine ORM issue (if it is an issue on our part).

**Comment by greg0ire** [comment]
Feel free to link to the new issue here though

**Comment by clytemnestra** [comment]
Should I open the same issue there?

**Comment by tim96** [comment]
> Closing because I can't see how this would not be a Sonata Doctrine ORM issue (if it is an issue on our part).

@greg0ire Can you explain the problem? I think we need simple add this moment to documentation. What do you think?

**Comment by tim96** [comment]
> I don't understand this part of the sentence.

About lazy load and join records. Because for me it's standard behavior.

**Comment by greg0ire** [comment]
> we need simple add this moment

Can you rephrase this though?

**Comment by tim96** [comment]
> Can you rephrase this though?

Create new PR to add some information about mappings and join records here:
https://sonata-project.org/bundles/admin/master/doc/reference/action_list.html#customizing-the-query-used-to-generate-the-list

**Comment by greg0ire** [comment]
> Create new PR to add some information about mappings and join records here:
https://sonata-project.org/bundles/admin/master/doc/reference/action_list.html#customizing-the-query-used-to-generate-the-list

I think it should be in the sonata doctrine **ORM** docs instead. What do you think ?

**Comment by tim96** [comment]
> What do you think ?

I'm not sure about it, because you work with Admin class so you will search information about it there.

**Comment by greg0ire** [comment]
> I'm not sure about it, because you work with Admin class so you will search information about it there.

That's right, but we cannot cram this documentation with persistence-backend specific things, can we? We could still add a link to the ORM docs though, at the end like this : 

`using doctrine ORM ? Learn more about listings here`

**Comment by tim96** [comment]
> That's right, but we cannot cram this documentation with persistence-backend specific things, can we? We could still add a link to the ORM docs though, at the end like this :

> using doctrine ORM ? Learn more about listings here

@greg0ire I think you are right.

**Comment by greg0ire** [comment]
Ok so now let's make real sure this is not supposed to happen. The query is created, but maybe it is supposed to be altered afterwards?

**Comment by greg0ire** [comment]
Can you open a new issue, and tell us if [this method](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/3.x/Datagrid/ProxyQuery.php#L242) is called? Looks like it does what you want.

**Comment by jjsmclaughlin** [comment]
I had the same problem. I found that it was solved by using a doctrine_orm_model_autocomplete as my requested Doctrine ORM filter type on the datagrid:

`$datagridMapper->add('my_foreign_key_property', 'doctrine_orm_model_autocomplete', [], null, ['property'=>'a_field_on_my_foreign_key_entity'])`

This seems to stop Doctrine from trying to populate the datagrid in advance with all those queries. In my case, it was tens of thousands.
